relay-server/src/actors/project_buffer.rs

```diff
@@ -113,66 +167,345 @@ impl FromMessage<RemoveMany> for Buffer {
+        if let Some(path) = config.cache_persistent_buffer_path() {
+            relay_log::info!("Using the buffer file: {}", path.to_string_lossy());
+
+            Self::setup(&path).await?;
+
+            let options = SqliteConnectOptions::new()
+                .filename(&path)
+                // The WAL journaling mode uses a write-ahead log instead of a rollback journal to implement transactions.
+                // The WAL journaling mode is persistent; after being set it stays in effect
+                // across multiple database connections and after closing and reopening the database.
+                //
+                // 1. WAL is significantly faster in most scenarios.
+                // 2. WAL provides more concurrency as readers do not block writers and a writer does not block readers. Reading and writing can proceed concurrently.
+                // 3. Disk I/O operations tends to be more sequential using WAL.
+                // 4. WAL uses many fewer fsync() operations and is thus less vulnerable to problems on systems where the fsync() system call is broken.
+                .journal_mode(SqliteJournalMode::Wal)
+                // If shared-cache mode is enabled and a thread establishes multiple
+                // connections to the same database, the connections share a single data and schema cache.
+                // This can significantly reduce the quantity of memory and IO required by the system.
+                .shared_cache(true);
+
+            let db = SqlitePoolOptions::new()
+                .max_connections(config.cache_persistent_buffer_max_connections())
+                .min_connections(config.cache_persistent_buffer_min_connections())
+                .connect_with(options)
+                .await?;
+
+            let spool_config = BufferSpoolConfig {
+                db,
+                max_disk_size: config.cache_persistent_buffer_max_disk_size(),
+                max_memory_size: config.cache_persistent_buffer_max_memory_size(),
+            };
+
+            service.spool_config = Some(spool_config);
         }
+
+        Ok(service)
+    }
+
+    /// Tries to save in-memory buffer to disk.
+    ///
+    /// It will spool to disk only if the persistent storage enabled in the configuration.
+    fn try_spool(&mut self) -> Result<(), BufferError> {
+        let Self {
+            config,
+            spool_config,
+            ref mut buffer,
+            ..
+        } = self;
+
+        // Buffer to disk only if the DB and config are provided.
+        if let Some(BufferSpoolConfig {
+            db,
+            max_disk_size,
+            max_memory_size,
+        }) = spool_config
+        {
+            // And if the count of in memory envelopes is over the defined max buffer size.
+            if self.count_mem_envelopes > *max_memory_size as i64 {
+                // Reject all the enqueue requests if we exceed the max size of the buffer.
```

Comment:
Question not necessarily related to the PR. If we have too many envelopes in memory and in the buffer, this actor will reject the incoming message. What happens with the envelope in that case? In case it's dropped, how can we prevent that?

Reply:
if we hit the limit, it will behave the way it is now, just start dropping the incoming envelopes.